File: dl-nn-in-many-languages/nn.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

#define N_EXAMPLES 150
#define N_EPOCH 1000
#define IN_DIM 4
#define OUT_DIM 3
#define HIDDEN_DIM 20
#define BS 30
#define LR 0.0003f

#define DATA_PATH "data/iris.data"
#define SPLIT_PATH "data/split.txt"
#define PARAMS_PATH "data/params.txt"

typedef struct params {
  float w1[IN_DIM][HIDDEN_DIM];
  float b1[HIDDEN_DIM];
  float w2[HIDDEN_DIM][OUT_DIM];
  float b2[OUT_DIM];
} params_t;

float random_uniform(float left, float right) {
  float a = rand() / (float)RAND_MAX;
  return (a + left) * (right - left);
}

void initialize_params(params_t *params) {
  float bound1 = sqrtf(6.0f / (IN_DIM + HIDDEN_DIM));
  for (size_t col = 0; col < HIDDEN_DIM; col++) {
    params->b1[col] = 0.0f;
    for (size_t row = 0; row < IN_DIM; row++) {
      params->w1[row][col] = random_uniform(-bound1, bound1);
    }
  }
  float bound2 = sqrtf(6.0f / (HIDDEN_DIM + OUT_DIM));
  for (size_t col = 0; col < OUT_DIM; col++) {
    params->b2[col] = 0.0f;
    for (size_t row = 0; row < HIDDEN_DIM; row++) {
      params->w2[row][col] = random_uniform(-bound2, bound2);
    }
  }
}

void load_params(params_t *params) {
  FILE *fp = fopen(PARAMS_PATH, "r");
  char line[1024];
  char *token;
  if (fgets(line, sizeof(line), fp)) {
    token = strtok(line, ",");
    for (size_t i = 0; i < IN_DIM; i++) {
      for (size_t j = 0; j < HIDDEN_DIM; j++) {
        params->w1[i][j] = atof(token);
        token = strtok(NULL, ",");
      }
    }
  }
  if (fgets(line, sizeof(line), fp)) {
    token = strtok(line, ",");
    for (size_t i = 0; i < HIDDEN_DIM; i++) {
      params->b1[i] = atof(token);
      token = strtok(NULL, ",");
    }
  }
  if (fgets(line, sizeof(line), fp)) {
    token = strtok(line, ",");
    for (size_t i = 0; i < HIDDEN_DIM; i++) {
      for (size_t j = 0; j < OUT_DIM; j++) {
        params->w2[i][j] = atof(token);
        token = strtok(NULL, ",");
      }
    }
  }
  if (fgets(line, sizeof(line), fp)) {
    token = strtok(line, ",");
    for (size_t i = 0; i < OUT_DIM; i++) {
      params->b2[i] = atof(token);
      token = strtok(NULL, ",");
    }
  }
  fclose(fp);
}

void matmul(size_t M, size_t N, size_t K, float A[M][K], float B[K][N], float ret[M][N]) {
  for (size_t m = 0; m < M; m++) {
    for (size_t n = 0; n < N; n++) {
      ret[m][n] = 0.0f;
      for (size_t k = 0; k < K; k++) {
        ret[m][n] += A[m][k] * B[k][n];
      }
    }
  }
}

float op_relu(float x) { return x > 0.0f ? x : 0.0f; }
float op_drelu(float g, float x) { return x > 0.0f ? g : 0.0f; }

void unary_elemwise(size_t M, size_t N, float A[M][N], float B[M][N], float (*op)(float)) {
  for (size_t i = 0; i < M; i++) {
    for (size_t j = 0; j < N; j++) {
      B[i][j] = op(A[i][j]);
    }
  }
}

void binary_elemwise(size_t M, size_t N, float A[M][N], float B[M][N], float C[M][N], float (*op)(float, float)) {
  for (size_t i = 0; i < M; i++) {
    for (size_t j = 0; j < N; j++) {
      C[i][j] = op(A[i][j], B[i][j]);
    }
  }
}

void transpose(size_t M, size_t N, float A[M][N], float B[N][M]) {
  for (size_t i = 0; i < M; i++) {
    for (size_t j = 0; j < N; j++) {
      B[j][i] = A[i][j];
    }
  }
}

void softmax(size_t M, size_t N, float A[M][N], float B[M][N]) {
  float expCache[N];
  for (size_t i = 0; i < M; i++) {
    float expSum = 0.0f;
    for (size_t j = 0; j < N; j++) {
      expCache[j] = expf(A[i][j]);
      expSum += expCache[j];
    }
    for (size_t j = 0; j < OUT_DIM; j++) {
      B[i][j] = expCache[j] / expSum;
    }
  }
}

void train_step(params_t* params, float feats[][IN_DIM], float labels[][OUT_DIM], size_t start) {
  /* forward */
  float h1[BS][HIDDEN_DIM];
  matmul(BS, HIDDEN_DIM, IN_DIM, &feats[start], params->w1, h1);
  for (size_t i = 0; i < BS; i++) {
    for (size_t j = 0; j < HIDDEN_DIM; j++) {
      h1[i][j] += params->b1[j];
    }
  }

  float a1[BS][HIDDEN_DIM];
  unary_elemwise(BS, HIDDEN_DIM, h1, a1, op_relu);

  float h2[BS][OUT_DIM];
  matmul(BS, OUT_DIM, HIDDEN_DIM, a1, params->w2, h2);
  for (size_t i = 0; i < BS; i++) {
    for (size_t j = 0; j < OUT_DIM; j++) {
      h2[i][j] += params->b2[j];
    }
  }

  float a2[BS][OUT_DIM];
  softmax(BS, OUT_DIM, h2, a2);

  /* NLL loss */
  /*
  float loss = 0.0f;
  for (size_t i = 0; i < BS; i++) {
    float nll = 0.0f;
    for (size_t j = 0; j < OUT_DIM; j++) {
      nll += a2[i][j] * labels[start + i][j];
    }
    loss += -logf(nll);
  }
  loss /= BS;
  printf("loss: %f\n", loss);
  */

  /* backward */
  float d_h2[BS][OUT_DIM];
  for (size_t i = 0; i < BS; i++) {
    for (size_t j = 0; j < OUT_DIM; j++) {
      d_h2[i][j] = a2[i][j] - labels[start + i][j];
    }
  }

  float a1_t[HIDDEN_DIM][BS];
  transpose(BS, HIDDEN_DIM, a1, a1_t);
  float d_w2[HIDDEN_DIM][OUT_DIM];
  matmul(HIDDEN_DIM, OUT_DIM, BS, a1_t, d_h2, d_w2);
  float d_b2[OUT_DIM] = {0.0f};
  for (size_t i = 0; i < BS; i++) {
    for (size_t j = 0; j < OUT_DIM; j++) {
      d_b2[j] += d_h2[i][j];
    }
  }

  float w2_t[OUT_DIM][HIDDEN_DIM];
  transpose(HIDDEN_DIM, OUT_DIM, params->w2, w2_t);
  float d_a1[BS][HIDDEN_DIM];
  matmul(BS, HIDDEN_DIM, OUT_DIM, d_h2, w2_t, d_a1);

  float d_h1[BS][HIDDEN_DIM];
  binary_elemwise(BS, HIDDEN_DIM, d_a1, h1, d_h1, op_drelu);

  float x_t[IN_DIM][BS];
  transpose(BS, IN_DIM, &feats[start], x_t);

  float d_w1[IN_DIM][HIDDEN_DIM];
  matmul(IN_DIM, HIDDEN_DIM, BS, x_t, d_h1, d_w1);

  float d_b1[HIDDEN_DIM] = {0.0f};
  for (size_t i = 0; i < BS; i++) {
    for (size_t j = 0; j < HIDDEN_DIM; j++) {
      d_b1[j] += d_h1[i][j];
    }
  }

  /* gradient descent */
  for (size_t i = 0; i < IN_DIM; i++) {
    for (size_t j = 0; j < HIDDEN_DIM; j++) {
      params->w1[i][j] -= LR * d_w1[i][j];
    }
  }
  for (size_t i = 0; i < HIDDEN_DIM; i++) {
    params->b1[i] -= LR * d_b1[i];
  }
  for (size_t i = 0; i < HIDDEN_DIM; i++) {
    for (size_t j = 0; j < OUT_DIM; j++) {
      params->w2[i][j] -= LR * d_w2[i][j];
    }
  }
  for (size_t i = 0; i < OUT_DIM; i++) {
    params->b2[i] -= LR * d_b2[i];
  }
}

void evaluate(params_t* params, float feats[][IN_DIM], float labels[][OUT_DIM], size_t start, size_t test_size) {
  /* forward */
  float h1[test_size][HIDDEN_DIM];
  matmul(BS, HIDDEN_DIM, IN_DIM, &feats[start], params->w1, h1);
  for (size_t i = 0; i < test_size; i++) {
    for (size_t j = 0; j < HIDDEN_DIM; j++) {
      h1[i][j] += params->b1[j];
    }
  }

  float a1[test_size][HIDDEN_DIM];
  unary_elemwise(test_size, HIDDEN_DIM, h1, a1, op_relu);
  float h2[test_size][OUT_DIM];
  matmul(test_size, OUT_DIM, HIDDEN_DIM, a1, params->w2, h2);
  for (size_t i = 0; i < test_size; i++) {
    for (size_t j = 0; j < OUT_DIM; j++) {
      h2[i][j] += params->b2[j];
    }
  }
  float a2[test_size][OUT_DIM];
  softmax(test_size, OUT_DIM, h2, a2);

  /* precision */
  size_t count = 0;
  for (size_t i = 0; i < test_size; i++) {
    size_t pred_label, true_label;
    float max_prob = -INFINITY;
    for (size_t j = 0; j < OUT_DIM; j++) {
      if (a2[i][j] > max_prob) {
        max_prob = a2[i][j];
        pred_label = j;
      }
      if (labels[start + i][j] == 1.0f) {
        true_label = j;
      }
    }
    if (pred_label == true_label) {
      count += 1;
    }
  }
  float precision = (float)count / test_size;


  float loss = 0.0f;
  for (size_t i = 0; i < test_size; i++) {
    float nll = 0.0f;
    for (size_t j = 0; j < OUT_DIM; j++) {
      nll += a2[i][j] * labels[start + i][j];
    }
    loss += -logf(nll);
  }
  loss /= test_size;
  printf("precision=%.6f, loss=%.6f\n", precision, loss);
}

void parse_example(char* line, float* feat, float* label) {
  const char *tok;
  size_t i = 0;
  for (tok = strtok(line, ","); tok; tok=strtok(NULL, ","), i++) {
    if (i < IN_DIM) {
      feat[i] = strtof(tok, NULL);
      continue;
    }
    if (strcmp(tok, "Iris-setosa") == 0) {
      label[0] = 1.0f;
    } else if (strcmp(tok, "Iris-versicolor") == 0) {
      label[1] = 1.0f;
    } else {
      label[2] = 1.0f;
    }
  }
}

void load_dataset(float feats[][IN_DIM], float labels[][OUT_DIM]) {
  FILE *fp = fopen(DATA_PATH, "r");
  char line[64];
  size_t i = 0;
  while (fgets(line, 64, fp)) {
    line[strcspn(line, "\n")] = '\0';    // get rid of the trailing \n
    if (strlen(line)) {
      char *dup = strdup(line);
      parse_example(dup, feats[i], labels[i]);
      free(dup);
      i++;
    }
  }
  fclose(fp);
}

void preprocess_dataset(float feats[][IN_DIM], float labels[][OUT_DIM]) {
  /* normalization */
  for (size_t j = 0; j < IN_DIM; j++) {
    float mean = 0.0f, std = 0.0f;
    for (size_t i = 0; i < N_EXAMPLES; i++) {
      mean += feats[i][j];
    }
    mean /= N_EXAMPLES;
    for (size_t i = 0; i < N_EXAMPLES; i++) {
      std += powf(feats[i][j] - mean, 2);
    }
    std = sqrtf(std / N_EXAMPLES);
    // printf("dim %zu mean: %f std: %f\n", j, mean, std);

    for (size_t i = 0; i < N_EXAMPLES; i++) {
      feats[i][j] = (feats[i][j] - mean) / std;
    }
  }

  /* shuffle dataset */
  FILE *fp = fopen(SPLIT_PATH, "r");
  char line[512];
  int random_indices[N_EXAMPLES];
  int count = 0;

  // Read indices
  if (fgets(line, sizeof(line), fp)) {
    char *token = strtok(line, ",");
    while (token) {
      random_indices[count++] = atoi(token);
      token = strtok(NULL, ",");
    }
  }

  if (fgets(line, sizeof(line), fp)) {
    char *token = strtok(line, ",");
    while (token) {
      random_indices[count++] = atoi(token);
      token = strtok(NULL, ",");
    }
  }
  fclose(fp);

  // temporary arrays to hold shuffled data
  float t_feats[N_EXAMPLES][IN_DIM];
  float t_labels[N_EXAMPLES][OUT_DIM];
  for (size_t i = 0; i < N_EXAMPLES; i++) {
    size_t j = random_indices[i];
    memcpy(t_feats[i], feats[j], sizeof(float) * IN_DIM);
    memcpy(t_labels[i], labels[j], sizeof(float) * OUT_DIM);
  }
  // shuffled data back to original arrays
  memcpy(feats, t_feats, sizeof(float) * N_EXAMPLES * IN_DIM);
  memcpy(labels, t_labels, sizeof(float) * N_EXAMPLES * OUT_DIM);
}

int main() {
  /* timing */
  clock_t start = clock();
  /* load data */
  int train_size = (int)(N_EXAMPLES * 0.8f);
  float feats[N_EXAMPLES][IN_DIM];
  float labels[N_EXAMPLES][OUT_DIM] = {{0.0f}};

  load_dataset(feats, labels);
  preprocess_dataset(feats, labels);

  params_t params;
  // initialize_params(&params);
  load_params(&params);
  for (int epoch = 0; epoch < N_EPOCH; epoch++) {
    for (int step = 0; step < train_size / BS; step++) {
      train_step(&params, feats, labels, step * BS);
    }
  }
  evaluate(&params, feats, labels, train_size, N_EXAMPLES - train_size);
  clock_t end = clock();
  printf("time=%.4fms\n", (float)(end - start) / CLOCKS_PER_SEC * 1000);
}

File: dl-nn-in-many-languages/nn.rs
use std::convert::TryInto;
use std::time::Instant;
use std::fs::File;
use std::io::{BufRead, BufReader};

const N_EPOCH: usize = 1000;
const N_EXAMPLES: usize = 150;
const IN_DIM: usize = 4;
const OUT_DIM: usize = 3;
const HIDDEN_DIM: usize = 20;
const BS: usize = 30;
const LR: f32 = 0.0003;

const PARAMS_PATH: &str = "data/params.txt";
const DATA_PATH: &str = "data/iris.data";
const SPLIT_PATH: &str = "data/split.txt";

#[derive(Debug)]
struct Params {
  w1: [[f32; HIDDEN_DIM]; IN_DIM],
  b1: [f32; HIDDEN_DIM],
  w2: [[f32; OUT_DIM]; HIDDEN_DIM],
  b2: [f32; OUT_DIM],
}

#[derive(Debug)]
struct Dataset {
  x: Vec<[f32; IN_DIM]>,
  y: Vec<[f32; OUT_DIM]>,
  _curr_idx: usize,
}

impl Dataset {
  pub fn new(x: Vec<[f32; IN_DIM]>, y: Vec<[f32; OUT_DIM]>) -> Self {
    Dataset { x, y, _curr_idx: 0 }
  }

  pub fn normalize(&mut self) {
    for j in 0..IN_DIM {
      // mean & std
      let mean: f32 = self.x.iter().map(|x| x[j]).sum::<f32>() / N_EXAMPLES as f32;
      let std: f32 = (self.x.iter().map(|x| (x[j] - mean).powi(2)).sum::<f32>() / N_EXAMPLES as f32).sqrt();
      // println!("dim {} mean: {} std: {}", j, mean, std);

      // Normalize the data
      for i in 0..N_EXAMPLES {
        self.x[i][j] = if std != 0.0 { (self.x[i][j] - mean) / std } else { 0.0 };
      }
    }
  }

  fn split(&self, split_path: &str) -> (Dataset, Dataset) {
    // read from split.txt
    let file = File::open(split_path).unwrap();
    let reader = BufReader::new(file);
    let lines: Vec<String> = reader.lines().filter_map(|line| line.ok()).collect();
    // ref: python/run_np.py
    let train_idx: Vec<usize> = lines[0].split(",").map(|s| s.parse::<usize>().unwrap()).collect();
    let test_idx: Vec<usize> = lines[1].split(",").map(|s| s.parse::<usize>().unwrap()).collect();

    let train_set = Dataset::new(
      train_idx.iter().map(|i| self.x[*i].clone()).collect(),
      train_idx.iter().map(|i| self.y[*i].clone()).collect(),
    );
    let test_set = Dataset::new(
      test_idx.iter().map(|i| self.x[*i].clone()).collect(),
      test_idx.iter().map(|i| self.y[*i].clone()).collect(),
    );
    (train_set, test_set)
  }

  fn size(&self) -> usize {
    self.x.len()
  }

  fn next_batch(&mut self) -> ([[f32; IN_DIM]; BS], [[f32; OUT_DIM]; BS]) {
    if self._curr_idx + BS > self.size() {
      self._curr_idx = 0;
    }
    let new_idx = self._curr_idx + BS;
    let x = self.x[self._curr_idx..new_idx].try_into().unwrap();
    let y = self.y[self._curr_idx..new_idx].try_into().unwrap();
    self._curr_idx = new_idx;
    (x, y)
  }

}

impl Params {
  fn new() -> Params {
      Params {
      w1: [[0.0; HIDDEN_DIM]; IN_DIM],
      b1: [0.0; HIDDEN_DIM],
      w2: [[0.0; OUT_DIM]; HIDDEN_DIM],
      b2: [0.0; OUT_DIM],
    }
  }

  fn from_file(&mut self, path: &str) {
    let file = File::open(path).unwrap();
    let reader = BufReader::new(file);
    let lines: Vec<String> = reader.lines().filter_map(|line| line.ok()).collect();
    let w1: Vec<f32> = lines[0].split(",").map(|s| s.parse::<f32>().unwrap()).collect();
    let b1: Vec<f32> = lines[1].split(",").map(|s| s.parse::<f32>().unwrap()).collect();
    let w2: Vec<f32> = lines[2].split(",").map(|s| s.parse::<f32>().unwrap()).collect();
    let b2: Vec<f32> = lines[3].split(",").map(|s| s.parse::<f32>().unwrap()).collect();

    // Reshape and assign to self
    for i in 0..IN_DIM {
      for j in 0..HIDDEN_DIM {
        self.w1[i][j] = w1[i * HIDDEN_DIM + j];
      }
    }
    self.b1.copy_from_slice(&b1);

    for i in 0..HIDDEN_DIM {
      for j in 0..OUT_DIM {
        self.w2[i][j] = w2[i * OUT_DIM + j];
      }
    }
    self.b2.copy_from_slice(&b2);
  }
}


fn load_dataset(path: &str, x_vec: &mut Vec<[f32; IN_DIM]>, y_vec: &mut Vec<[f32; OUT_DIM]>) {
  let file = File::open(path).unwrap();
  let reader = BufReader::new(file);

  for line in reader.lines() {
    let record = line.unwrap();
    let fields: Vec<&str> = record.split(",").collect();

    let feats_vec: Vec<f32> = fields.iter().take(IN_DIM).map(|s| s.parse::<f32>().unwrap()).collect();
    let x: [f32; IN_DIM] = feats_vec.try_into().expect("Incorrect length for feat array");

    let label = fields[IN_DIM];
    let y: [f32; OUT_DIM] = match label {
        "Iris-setosa" => [1.0, 0.0, 0.0],
        "Iris-versicolor" => [0.0, 1.0, 0.0],
        "Iris-virginica" => [0.0, 0.0, 1.0],
        _ => panic!("Invalid label value '{}'", label),
    };
    x_vec.push(x);
    y_vec.push(y);
  }
}

fn matmul<const M: usize, const N: usize, const K: usize>(
  a: &[[f32; K]; M], b: &[[f32; N]; K]
) -> [[f32; N]; M] {
  let mut ret:[[f32; N]; M] = [[0.0; N]; M];
  for i in 0..M {
    for j in 0..N {
      ret[i][j] = 0.0;
      for k in 0..K {
        ret[i][j] += a[i][k] * b[k][j];
      }
    }
  }
  ret
}

fn softmax<const M: usize, const N: usize> (input: &[[f32; N]; M]) -> [[f32; N]; M] {
  let mut ret: [[f32; N]; M] = [[0.0; N]; M];
  let mut exp_cache: [f32; N] = [0.0; N];
  for i in 0..M {
    let mut exp_sum: f32 = 0.0;
    for j in 0..N {
      exp_cache[j] = input[i][j].exp();
      exp_sum += exp_cache[j];
    }
    for j in 0..N {
      ret[i][j] = exp_cache[j] / exp_sum;
    }
  }
  ret
}

fn unary_elemwise<F, const M: usize, const N: usize> (
  input: &[[f32; N]; M], op: F,
) -> [[f32; N]; M] where F: Fn(f32) -> f32 {
  let mut ret = [[0.0; N]; M];
  for i in 0..M {
    for j in 0..N {
      ret[i][j] = op(input[i][j]);
    }
  }
  ret
}

fn binary_elemwise<F, const M: usize, const N: usize> (
  a: &[[f32; N]; M], b: &[[f32; N]; M], op: F,
) -> [[f32; N]; M] where F: Fn(f32, f32) -> f32 {
  let mut ret = [[0.0; N]; M];
  for i in 0..M {
    for j in 0..N {
      ret[i][j] = op(a[i][j], b[i][j]);
    }
  }
  ret
}

fn transpose<const M: usize, const N: usize> (
  input: &[[f32; N]; M]
) -> [[f32; M]; N] {
  let mut ret = [[0.0; M]; N];
  for i in 0..M {
    for j in 0..N {
      ret[j][i] = input[i][j];
    }
  }
  ret
}

fn train_step(params: &mut Params, x: &[[f32; IN_DIM]; BS], y: &[[f32; OUT_DIM]; BS]) {
  // Forward pass
  let mut h1 = matmul(x, &params.w1);
  for i in 0..BS {
    for j in 0..HIDDEN_DIM {
      h1[i][j] += params.b1[j];
    }
  }
  let a1 = unary_elemwise(&h1, |x| if x > 0.0 {x} else {0.0});

  let mut h2 = matmul(&a1, &params.w2);
  for i in 0..BS {
    for j in 0..OUT_DIM {
      h2[i][j] += params.b2[j];
    }
  }
  let a2 = softmax(&mut h2);

  // // Loss calculation: Cross-entropy loss
  // let mut loss: f32 = 0.0;
  // for i in 0..BS {
  //   let mut nll: f32 = 0.0;
  //   for j in 0..OUT_DIM {
  //     nll += a2[i][j] * y[i][j];
  //   }
  //   loss += -nll.ln();
  // }
  // loss /= BS as f32;
  // println!("loss: {}", loss);

  // Backward pass
  // d_h2 = a2 - y (derivative of softmax with cross-entropy)
  let d_h2 = binary_elemwise(&a2, &y, |a, b| a - b);
  // d_w2 = a1^T * d_h2
  let d_w2 = matmul(&transpose(&a1), &d_h2);
  // d_b2 = sum(d_h2, axis=0)
  let mut d_b2 = [0.0; OUT_DIM];
  for i in 0..BS {
    for j in 0..OUT_DIM {
      d_b2[j] += d_h2[i][j];
    }
  }

  // d_a1 = d_h2 * w2^T
  let d_a1 = matmul(&d_h2, &transpose(&params.w2));
  // d_h1 = d_a1 * ReLU'(h1)
  let d_h1 = binary_elemwise(&d_a1, &h1, |a, b| if b > 0.0 {a} else {0.0});

  // d_w1 = x^T * d_h1
  let d_w1 = matmul(&transpose(x), &d_h1);
  // d_b1 = sum(d_h1, axis=0)
  let mut d_b1 = [0.0; HIDDEN_DIM];
  for i in 0..BS {
    for j in 0..HIDDEN_DIM {
      d_b1[j] += d_h1[i][j];
    }
  }

  // Gradient descent update
  for i in 0..IN_DIM {
    for j in 0..HIDDEN_DIM {
      params.w1[i][j] -= LR * d_w1[i][j];
    }
  }
  for i in 0..HIDDEN_DIM {
    params.b1[i] -= LR * d_b1[i];
  }
  for i in 0..HIDDEN_DIM {
    for j in 0..OUT_DIM {
      params.w2[i][j] -= LR * d_w2[i][j];
    }
  }
  for i in 0..OUT_DIM {
    params.b2[i] -= LR * d_b2[i];
  }
}

fn evaluate(params: &Params, x: &[[f32; IN_DIM]; BS], y: &[[f32; OUT_DIM]; BS]) -> (f32, f32) {
  // forward pass and calcualte precision
  let mut h1 = matmul(x, &params.w1);
  for i in 0..BS {
    for j in 0..HIDDEN_DIM {
      h1[i][j] += params.b1[j];
    }
  }
  let a1 = unary_elemwise(&h1, |x| if x > 0.0 {x} else {0.0});
  let mut h2 = matmul(&a1, &params.w2);
  for i in 0..BS {
    for j in 0..OUT_DIM {
      h2[i][j] += params.b2[j];
    }
  }
  let a2 = softmax(&mut h2);

  let mut count = 0;
  for i in 0..BS {
    let mut max_prob = -f32::INFINITY;
    let (mut pred_label, mut true_label) = (0, 0);
    for j in 0..OUT_DIM {
      if a2[i][j] > max_prob {
        max_prob = a2[i][j];
        pred_label = j;
      }
      if y[i][j] == 1.0 {
        true_label = j;
      }
    }
    if pred_label == true_label {
      count += 1;
    }
  }
  let precision = count as f32 / BS as f32;
  let mut loss: f32 = 0.0;
  for i in 0..BS {
    let mut nll: f32 = 0.0;
    for j in 0..OUT_DIM {
      nll += a2[i][j] * y[i][j];
    }
    loss += -nll.ln();
  }
  loss /= BS as f32;

  (precision, loss)
}

fn main() {
  // timing
  let st = Instant::now();

  let mut x: Vec<[f32; IN_DIM]> = Vec::new();
  let mut y: Vec<[f32; OUT_DIM]> = Vec::new();
  load_dataset(DATA_PATH, &mut x, &mut y);

  let mut dataset = Dataset::new(x, y);
  dataset.normalize();
  let (mut train_set, test_set) = dataset.split(SPLIT_PATH);

  // init params
  let mut params = Params::new();
  params.from_file(PARAMS_PATH);

  // train & evaluate
  let test_x: [[f32; IN_DIM]; BS] = test_set.x.clone().try_into().unwrap();
  let test_y: [[f32; OUT_DIM]; BS] = test_set.y.clone().try_into().unwrap();

  for _epoch in 0..N_EPOCH {
    for _step in 0..(train_set.size() / BS) {
      let (x, y) = train_set.next_batch();
      train_step(&mut params, &x, &y);
    }
  }
  let (precision, loss) = evaluate(&params, &test_x, &test_y);
  println!("precision={:.6}, loss={:.6}", precision, loss);
  let duration = st.elapsed();
  println!("time={:.4}ms", duration.as_secs_f32() * 1000.0);
}
